File: public_samples/quickstart/text-to-speech/objectivec-ios/helloworld/helloworld/ViewController.m
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

#import "ViewController.h"
#import <AVFoundation/AVFoundation.h>
#import <MicrosoftCognitiveServicesSpeech/SPXSpeechApi.h>

@interface ViewController () {
    NSString *inputText;
    
    NSString *speechKey;
    NSString *serviceRegion;
}

@property (nonatomic, strong) AVAudioPlayer *player;
- (IBAction)synthesisButtonTapped:(UIButton *)sender;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    inputText = @"";
    
    speechKey = @"YourSubscriptionKey";
    serviceRegion = @"YourServiceRegion";
}

- (IBAction)synthesisButtonTapped:(UIButton *)sender {
    dispatch_async(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
        [self synthesis];
    });
}

- (void)synthesis {
    SPXSpeechConfiguration *speechConfig = [[SPXSpeechConfiguration alloc] initWithSubscription:speechKey region:serviceRegion];
    [speechConfig setSpeechSynthesisOutputFormat:SPXSpeechSynthesisOutputFormat_Audio16Khz32KBitRateMonoMp3];
    SPXSpeechSynthesizer *speechSynthesizer = [[SPXSpeechSynthesizer alloc] initWithSpeechConfiguration:speechConfig audioConfiguration:nil];
    
    NSLog(@"Start synthesizing...");
    
    SPXSpeechSynthesisResult *speechResult = [speechSynthesizer SpeakText:inputText];
    
    // Checks result.
    if (SPXResultReason_Canceled == speechResult.reason) {
        SPXSpeechSynthesisCancellationDetails *details = [[SPXSpeechSynthesisCancellationDetails alloc] initFromCanceledSynthesisResult:speechResult];
        NSLog(@"Speech synthesis was canceled: %@. Did you pass the correct key/region combination?", details.errorDetails);
    } else if (SPXResultReason_SynthesizingAudioCompleted == speechResult.reason) {
        NSLog(@"Speech synthesis was completed");
        // Play audio.
        self.player = [[AVAudioPlayer alloc] initWithData:[speechResult audioData] error:nil];
        [self.player prepareToPlay];
        [self.player play];
    } else {
        NSLog(@"There was an error.");
    }
}

- (BOOL)textField:(UITextField *)textField shouldChangeCharactersInRange:(NSRange)range replacementString:(NSString *)string {;
    inputText = [textField.text stringByReplacingCharactersInRange:range withString:string];
    return true;
}

@end

File: public_samples/quickstart/text-to-speech/swift-ios/helloworld/helloworld/ViewController.swift
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

// <code>
import UIKit
import AVFoundation

var player: AVAudioPlayer?

class ViewController: UIViewController, UITextFieldDelegate, AVAudioPlayerDelegate{
    var textField: UITextField!
    var synthesisButton: UIButton!
    
    var inputText: String!
    
    var sub: String!
    var region: String!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        // load subscription information
        sub = "YourSubscriptionKey"
        region = "YourServiceRegion"
        
        textField = UITextField(frame: CGRect(x: 100, y:250, width: 200, height: 50))
        textField.textColor = UIColor.black
        textField.borderStyle = UITextField.BorderStyle.roundedRect
        textField.placeholder = "Type something to synthesize."
        textField.delegate = self
        
        inputText = ""
        
        synthesisButton = UIButton(frame: CGRect(x: 100, y: 400, width: 200, height: 50))
        synthesisButton.setTitle("Synthesize", for: .normal)
        synthesisButton.addTarget(self, action:#selector(self.synthesisButtonClicked), for: .touchUpInside)
        synthesisButton.setTitleColor(UIColor.black, for: .normal)
        
        self.view.addSubview(textField)
        self.view.addSubview(synthesisButton)
    }
    
    func textField(_ textField: UITextField, shouldChangeCharactersIn range: NSRange, replacementString string: String) -> Bool {
        if let text = textField.text,
        let textRange = Range(range, in: text) {
            self.inputText = text.replacingCharacters(in: textRange, with: string)
        }
        return true
    }
    
    @objc func synthesisButtonClicked() {
        DispatchQueue.global(qos: .userInitiated).async {
            self.synthesisToSpeaker()
        }
    }
    
    func synthesisToSpeaker() {
        var speechConfig: SPXSpeechConfiguration?
        do {
            try speechConfig = SPXSpeechConfiguration(subscription: sub, region: region)
        } catch {
            print("error \(error) happened")
            speechConfig = nil
        }
        speechConfig?.setSpeechSynthesisOutputFormat( SPXSpeechSynthesisOutputFormat.audio16Khz32KBitRateMonoMp3)
        let synthesizer = try! SPXSpeechSynthesizer(speechConfiguration: speechConfig!, audioConfiguration: nil)
        if inputText.isEmpty {
            return
        }
        let result = try! synthesizer.speakText(inputText)
        do {
            try player = AVAudioPlayer(data: result.audioData!)
            player?.delegate = self
            player?.prepareToPlay()
        } catch {
            print("error \(error) happened")
            player = nil
        }
        player?.delegate = self
        player?.play()
    }
}
// </code>
